sound/soc/sof/ipc4-topology.c

```diff
@@ -1399,6 +1479,86 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_sof_control *scontrol = NULL;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	struct sof_ipc4_control_data *control_data;
+	struct sof_ipc4_available_audio_format *available_fmt = &process->available_fmt;
+	void *cfg = process->ipc_config_data;
+	void *extra_flag;
+	int ret, i;
+
+	available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules setup the output format based on audio format tokens in topology.
+	 */
+	if (swidget->output_fmt_cnt)
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      &process->output_format, pipeline_params,
+					      available_fmt, sizeof(process->base_config));
+	else
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      NULL, pipeline_params, available_fmt,
+					      sizeof(process->base_config));
+	if (ret < 0)
+		return ret;
+
+	/* update pipeline memory usage */
+	sof_ipc4_update_pipeline_mem_usage(sdev, swidget, &process->base_config);
+
+	/*
+	 * ipc_config_data is composed of the base_config, optional output formats followed
+	 * by the data required for module init in that order.
+	 */
+	memcpy(cfg, &process->base_config, sizeof(process->base_config));
+	cfg += sizeof(process->base_config);
+
+	/* copy output_format to configure data payload */
+	for (i = 0; i < swidget->output_fmt_cnt; i++) {
+		/* TODO: to handle different output formats */
+		memcpy(cfg, &process->output_format, sizeof(struct sof_ipc4_audio_format));
```

Comment:
this would mean you copy the same output format swidget->output_fmt_cnt times? Is this intentional?

Reply:
Please see the comment https://github.com/thesofproject/linux/pull/3755#discussion_r976325376

Reply:
Although output_fmt_cnt is an u32 and we want it to be extendable,  I think we should use it as bool, because we don't have module that has more than one out_fmt in its init blob. The module that has out_fmt is even rare case.

Reply:
Either assume in every place that output_fmt_cnt can be 0 or 1 (in which case it is not a count, but a module has output or something) or handle it as a counter.
But the mixed use of it is confusing.